File: src/main/java/org/example/kalkulationsprogramm/domain/LieferantGeschaeftsdokument.java
package org.example.kalkulationsprogramm.domain;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Geschäftsmetadaten für Lieferanten-Dokumente.
 * Erweitert LieferantDokument um strukturierte Daten wie Dokumentnummer,
 * Beträge, Liefertermin etc.
 * Diese Daten werden primär durch KI-Analyse extrahiert.
 */
@Getter
@Setter
@Entity
@Table(name = "lieferant_geschaeftsdokument")
public class LieferantGeschaeftsdokument {

    @Id
    private Long id; // = LieferantDokument.id

    @OneToOne(fetch = FetchType.LAZY)
    @MapsId
    @JoinColumn(name = "id")
    private LieferantDokument dokument;

    // Basis-Metadaten
    @Column(length = 50)
    private String dokumentNummer; // z.B. RE-2024-001

    private LocalDate dokumentDatum;

    @Column(precision = 12, scale = 2)
    private BigDecimal betragNetto;

    @Column(precision = 12, scale = 2)
    private BigDecimal betragBrutto;

    @Column(precision = 5, scale = 4)
    private BigDecimal mwstSatz; // z.B. 0.19 für 19%

    // Lieferanten-spezifische Felder
    private LocalDate liefertermin;

    @Column(length = 50)
    private String bestellnummer; // Unsere Bestellnummer beim Lieferanten

    @Column(length = 50)
    private String referenzNummer; // Für Verknüpfung (z.B. AB-Nr auf Rechnung)

    // KI-Analyse Metadaten
    @Column(columnDefinition = "TEXT")
    private String aiRawJson; // Original JSON-Response der KI

    private Double aiConfidence; // Vertrauenswert 0.0 - 1.0

    private LocalDateTime analysiertAm;

    // Zahlungsstatus für Offene Posten
    private LocalDate zahlungsziel; // Fälligkeitsdatum

    @Column(nullable = false)
    private Boolean bezahlt = false; // true = bezahlt

    private LocalDate bezahltAm; // Wann wurde bezahlt

    private Boolean bereitsGezahlt = false; // Von KI erkannt (Vorauskasse etc.)

    // Skonto-Konditionen (z.B. "8 TAGE 2% 30 TAGE NETTO")
    private Integer skontoTage; // Tage ab Rechnungsdatum für Skonto

    @Column(precision = 5, scale = 2)
    private BigDecimal skontoProzent; // z.B. 2.00 für 2%

    private Integer nettoTage; // Zahlungsziel in Tagen (z.B. 30)

    // Tatsächlich gezahlter Betrag (nach Skonto-Abzug)
    @Column(precision = 12, scale = 2)
    private BigDecimal tatsaechlichGezahlt;

    private Boolean mitSkonto = false; // true wenn Skonto genutzt wurde

    // Lagerbestellung-Flag (Bestellung ohne Projekt-Zuordnung)
    @Column(nullable = false)
    private Boolean lagerbestellung = false; // true = Lagerbestellung, keine Projekt-Zuordnung nötig

    private Boolean verifiziert = false; // true wenn strukturiert aus ZUGFeRD/XML
    private String datenquelle; // ZUGFERD, XML, AI

    @Column(nullable = false)
    private Boolean genehmigt = false; // Genehmigung durch Abteilung 3 (Büro)

    /**
     * Flag für manuelle Prüfung erforderlich.
     * Wird gesetzt wenn die KI-Extraktion unvollständig war (Pflichtfelder fehlen).
     * Der User muss das Dokument manuell nachbearbeiten.
     */
    @Column(nullable = false)
    private Boolean manuellePruefungErforderlich = false;

    // Transient field to transport detected type from AI analysis to creation
    // service
    @Transient
    private LieferantDokumentTyp detectedTyp;
}

File: src/main/kotlin/org/example/kalkulationsprogramm/domain/LieferantGeschaeftsdokument.kt
package org.example.kalkulationsprogramm.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.MapsId
import jakarta.persistence.OneToOne
import jakarta.persistence.Table
import jakarta.persistence.Transient
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime

@Entity
@Table(name = "lieferant_geschaeftsdokument")
open class LieferantGeschaeftsdokument {
    @Id
    open var id: Long? = null

    @OneToOne(fetch = FetchType.LAZY)
    @MapsId
    @JoinColumn(name = "id")
    open var dokument: LieferantDokument? = null

    @Column(length = 50)
    open var dokumentNummer: String? = null

    open var dokumentDatum: LocalDate? = null

    @Column(precision = 12, scale = 2)
    open var betragNetto: BigDecimal? = null

    @Column(precision = 12, scale = 2)
    open var betragBrutto: BigDecimal? = null

    @Column(precision = 5, scale = 4)
    open var mwstSatz: BigDecimal? = null

    open var liefertermin: LocalDate? = null

    @Column(length = 50)
    open var bestellnummer: String? = null

    @Column(length = 50)
    open var referenzNummer: String? = null

    @Column(columnDefinition = "TEXT")
    open var aiRawJson: String? = null

    open var aiConfidence: Double? = null
    open var analysiertAm: LocalDateTime? = null
    open var zahlungsziel: LocalDate? = null

    @Column(nullable = false)
    open var bezahlt: Boolean? = false

    open var bezahltAm: LocalDate? = null
    open var bereitsGezahlt: Boolean? = false

    @Column(length = 50)
    open var zahlungsart: String? = null

    open var skontoTage: Int? = null

    @Column(precision = 5, scale = 2)
    open var skontoProzent: BigDecimal? = null

    open var nettoTage: Int? = null

    @Column(precision = 12, scale = 2)
    open var tatsaechlichGezahlt: BigDecimal? = null

    open var mitSkonto: Boolean? = false

    @Column(nullable = false)
    open var lagerbestellung: Boolean? = false

    open var verifiziert: Boolean? = false
    open var datenquelle: String? = null

    @Column(nullable = false)
    open var genehmigt: Boolean? = false

    @Column(nullable = false)
    open var manuellePruefungErforderlich: Boolean? = false

    @Transient
    open var detectedTyp: LieferantDokumentTyp? = null

    open fun isBezahlt(): Boolean? = bezahlt
}
